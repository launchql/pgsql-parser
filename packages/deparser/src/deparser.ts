import { Node } from '@pgsql/types';
import { SqlFormatter } from './utils/sql-formatter';
import { ConsolidatedVisitor } from './visitors/consolidated-visitor';
import { DeparserContext } from './visitors/base';

export interface DeparserOptions {
  newline?: string;
  tab?: string;
}

export class Deparser {
  private formatter: SqlFormatter;
  private consolidatedVisitor: ConsolidatedVisitor;
  private tree: Node[];

  constructor(tree: Node | Node[], opts: DeparserOptions = {}) {
    this.formatter = new SqlFormatter(opts.newline, opts.tab);
    this.consolidatedVisitor = new ConsolidatedVisitor(this.formatter, this);
    
    this.tree = Array.isArray(tree) ? tree : [tree];
  }

  static deparse(query: Node | Node[], opts: DeparserOptions = {}): string {
    return new Deparser(query, opts).deparseQuery();
  }

  deparseQuery(): string {
    return this.tree
      .map(node => this.deparse(node))
      .join(this.formatter.newline() + this.formatter.newline());
  }

  deparse(node: Node, context: DeparserContext = {}): string | null {
    if (node == null) {
      return null;
    }

    if (typeof node === 'number' || node instanceof Number) {
      return node.toString();
    }

    try {
      return this.consolidatedVisitor.visit(node, context);
    } catch (error) {
      const nodeType = Object.keys(node)[0];
      throw new Error(`Error deparsing ${nodeType}: ${(error as Error).message}`);
    }
  }


}
