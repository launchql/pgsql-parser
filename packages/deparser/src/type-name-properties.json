{
  "TypeCast": ["typeName"],
  "RangeTableFuncCol": ["typeName"],
  "ColumnDef": ["typeName"],
  "XmlSerialize": ["typeName"],
  "JsonOutput": ["typeName"],
  "JsonTableColumn": ["typeName"],
  "CreateDomainStmt": ["typeName"],
  "CreateTransformStmt": ["type_name"],
  "CreateStmt": ["ofTypename"],
  "CreateOpClassItem": ["datatype", "storedtype"],
  "CreateFunctionStmt": ["returnType"],
  "FunctionParameter": ["argType"],
  "CreateCastStmt": ["sourcetype", "targettype"]
} 