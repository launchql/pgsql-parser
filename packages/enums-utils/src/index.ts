import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
