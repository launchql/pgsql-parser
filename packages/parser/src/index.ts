import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
