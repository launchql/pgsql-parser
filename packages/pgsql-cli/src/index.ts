#!/usr/bin/env node
import minimist from 'minimist';
import chalk from 'chalk';
import { readFileSync } from 'fs';
import { resolve } from 'path';

import { parseCommand } from './commands/parse';
import { deparseCommand } from './commands/deparse';
import { protoGenCommand } from './commands/proto-gen';
import { protoFetchCommand } from './commands/proto-fetch';
import { runtimeSchemaCommand } from './commands/runtime-schema';
import { showHelp, showVersion } from './utils/help';

const argv = minimist(process.argv.slice(2), {
  alias: {
    h: 'help',
    v: 'version',
    o: 'output',
    f: 'format'
  }
});

const command = argv._[0];

async function main() {
  try {
    if (argv.version) {
      showVersion();
      process.exit(0);
    }

    if (!command || argv.help) {
      showHelp(command);
      process.exit(0);
    }

    switch (command) {
      case 'parse':
        await parseCommand(argv);
        break;
      
      case 'deparse':
        await deparseCommand(argv);
        break;
      
      case 'proto-gen':
        await protoGenCommand(argv);
        break;
      
      case 'proto-fetch':
        await protoFetchCommand(argv);
        break;
      
      case 'runtime-schema':
        await runtimeSchemaCommand(argv);
        break;
      
      default:
        console.error(chalk.red(`Unknown command: ${command}`));
        showHelp();
        process.exit(1);
    }
  } catch (error: any) {
    console.error(chalk.red('Error:'), error.message || error);
    process.exit(1);
  }
}

main();
