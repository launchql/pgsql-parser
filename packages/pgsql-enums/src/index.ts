import toInt from './enums2int';
import toStr from './enums2str';
import nodes from './nodes';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
