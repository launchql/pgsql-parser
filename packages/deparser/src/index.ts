import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
