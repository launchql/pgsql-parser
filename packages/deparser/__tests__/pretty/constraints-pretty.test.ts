import { deparseSync } from '../../src';
import { parse } from 'libpg-query';

describe('Pretty constraint formatting', () => {
  const foreignKeyConstraintSql = `ALTER TABLE products ADD CONSTRAINT fk_category FOREIGN KEY (category_id) REFERENCES categories(id) ON UPDATE CASCADE ON DELETE SET NULL DEFERRABLE INITIALLY DEFERRED;`;
  
  const checkConstraintSql = `ALTER TABLE products ADD CONSTRAINT check_price CHECK (price > 0);`;

  const complexTableSql = `CREATE TABLE orders (
    id SERIAL PRIMARY KEY,
    user_id INTEGER NOT NULL,
    total DECIMAL(10,2) CHECK (total > 0),
    status VARCHAR(20) DEFAULT 'pending',
    CONSTRAINT fk_user FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
  );`;

  it('should format foreign key constraint with pretty option enabled', async () => {
    const parsed = await parse(foreignKeyConstraintSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format when pretty option disabled', async () => {
    const parsed = await parse(foreignKeyConstraintSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format check constraint with pretty option enabled', async () => {
    const parsed = await parse(checkConstraintSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format complex table with constraints with pretty option enabled', async () => {
    const parsed = await parse(complexTableSql);
    const result = deparseSync(parsed, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should maintain single-line format for complex table when pretty disabled', async () => {
    const parsed = await parse(complexTableSql);
    const result = deparseSync(parsed, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should use custom newline and tab characters in pretty mode', async () => {
    const parsed = await parse(foreignKeyConstraintSql);
    const result = deparseSync(parsed, { 
      pretty: true, 
      newline: '\r\n', 
      tab: '    ' 
    });
    expect(result).toMatchSnapshot();
  });
});
