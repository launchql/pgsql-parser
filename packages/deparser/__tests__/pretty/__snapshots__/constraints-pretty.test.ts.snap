// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`non-pretty: pretty/constraints-1.sql 1`] = `"ALTER TABLE public.users ADD CONSTRAINT users_pkey PRIMARY KEY (id)"`;

exports[`non-pretty: pretty/constraints-2.sql 1`] = `"ALTER TABLE "App"."User Data" ADD CONSTRAINT "Unique_Full Name" UNIQUE ("Full Name")"`;

exports[`non-pretty: pretty/constraints-3.sql 1`] = `"ALTER TABLE school.attendance ADD CONSTRAINT attendance_unique UNIQUE ("Student ID", "Class ID")"`;

exports[`non-pretty: pretty/constraints-4.sql 1`] = `"ALTER TABLE "Orders"."OrderLines" ADD CONSTRAINT "FK_Order_Ref" FOREIGN KEY (order_id) REFERENCES "Orders"."Order" ("OrderID")"`;

exports[`non-pretty: pretty/constraints-5.sql 1`] = `"ALTER TABLE "x-Schema"."z-Table" ADD CONSTRAINT "zNameFormatCheck" CHECK ("Z-Name" ~ '^[A-Z]')"`;

exports[`non-pretty: pretty/constraints-6.sql 1`] = `"ALTER TABLE data.snapshots ADD CONSTRAINT metadata_has_key CHECK (metadata ? 'type')"`;

exports[`non-pretty: pretty/constraints-7.sql 1`] = `"ALTER TABLE "Billing"."Invoices" ADD CONSTRAINT "FK_Client_ID" FOREIGN KEY ("Client ID") REFERENCES "Clients"."ClientBase" ("Client ID")"`;

exports[`non-pretty: pretty/constraints-8.sql 1`] = `"ALTER TABLE "API Keys" ADD CONSTRAINT "PK_KeyID" PRIMARY KEY ("KeyID")"`;

exports[`non-pretty: pretty/constraints-9.sql 1`] = `"ALTER TABLE finance.transactions ADD CONSTRAINT tax_rate_range CHECK (tax_rate >= 0 AND tax_rate <= 1)"`;

exports[`non-pretty: pretty/constraints-10.sql 1`] = `"ALTER TABLE school.enrollments ADD CONSTRAINT fk_student_course FOREIGN KEY (student_id, course_id) REFERENCES school.courses_students (student_id, course_id)"`;

exports[`non-pretty: pretty/constraints-11.sql 1`] = `"CREATE TABLE orders (id serial PRIMARY KEY, user_id int NOT NULL, total numeric(10, 2) CHECK (total > 0), status varchar(20) DEFAULT 'pending', created_at timestamp DEFAULT now(), CONSTRAINT fk_user FOREIGN KEY (user_id) REFERENCES users (id) ON DELETE CASCADE, CONSTRAINT unique_user_date UNIQUE (user_id, created_at), CONSTRAINT check_status CHECK (status IN ('pending', 'completed', 'cancelled')))"`;

exports[`non-pretty: pretty/constraints-12.sql 1`] = `"ALTER TABLE products ADD CONSTRAINT fk_category FOREIGN KEY (category_id) REFERENCES categories (id) ON UPDATE CASCADE ON DELETE SET NULL DEFERRABLE INITIALLY DEFERRED"`;

exports[`non-pretty: pretty/constraints-13.sql 1`] = `"ALTER TABLE products ADD CONSTRAINT check_price CHECK (price > 0)"`;

exports[`non-pretty: pretty/constraints-14.sql 1`] = `"ALTER TABLE users ADD CONSTRAINT unique_email UNIQUE (email)"`;

exports[`non-pretty: pretty/constraints-15.sql 1`] = `"ALTER TABLE school.enrollments ADD CONSTRAINT fk_student_course FOREIGN KEY (student_id, course_id) REFERENCES school.courses_students (student_id, course_id)"`;

exports[`non-pretty: pretty/constraints-16.sql 1`] = `"ALTER TABLE school.enrollments ADD CONSTRAINT chk_enrollment_date CHECK (enrollment_date <= CURRENT_DATE AND status IN ('active', 'completed', 'withdrawn'))"`;

exports[`non-pretty: pretty/constraints-17.sql 1`] = `"CREATE TABLE school.enrollments (student_id int NOT NULL, course_id int NOT NULL, enrollment_date date NOT NULL, status text CHECK (status IN ('active', 'completed', 'withdrawn')), CHECK (enrollment_date <= CURRENT_DATE))"`;

exports[`pretty: pretty/constraints-1.sql 1`] = `
"ALTER TABLE public.users 
  ADD CONSTRAINT users_pkey PRIMARY KEY (id)"
`;

exports[`pretty: pretty/constraints-2.sql 1`] = `
"ALTER TABLE "App"."User Data" 
  ADD CONSTRAINT "Unique_Full Name" 
    UNIQUE ("Full Name")"
`;

exports[`pretty: pretty/constraints-3.sql 1`] = `
"ALTER TABLE school.attendance 
  ADD CONSTRAINT attendance_unique 
    UNIQUE ("Student ID", "Class ID")"
`;

exports[`pretty: pretty/constraints-4.sql 1`] = `
"ALTER TABLE "Orders"."OrderLines" 
  ADD CONSTRAINT "FK_Order_Ref"
    FOREIGN KEY(order_id)
    REFERENCES "Orders"."Order" ("OrderID")"
`;

exports[`pretty: pretty/constraints-5.sql 1`] = `
"ALTER TABLE "x-Schema"."z-Table" 
  ADD CONSTRAINT "zNameFormatCheck" 
    CHECK ("Z-Name" ~ '^[A-Z]')"
`;

exports[`pretty: pretty/constraints-6.sql 1`] = `
"ALTER TABLE data.snapshots 
  ADD CONSTRAINT metadata_has_key 
    CHECK (metadata ? 'type')"
`;

exports[`pretty: pretty/constraints-7.sql 1`] = `
"ALTER TABLE "Billing"."Invoices" 
  ADD CONSTRAINT "FK_Client_ID"
    FOREIGN KEY("Client ID")
    REFERENCES "Clients"."ClientBase" ("Client ID")"
`;

exports[`pretty: pretty/constraints-8.sql 1`] = `
"ALTER TABLE "API Keys" 
  ADD CONSTRAINT "PK_KeyID" PRIMARY KEY ("KeyID")"
`;

exports[`pretty: pretty/constraints-9.sql 1`] = `
"ALTER TABLE finance.transactions 
  ADD CONSTRAINT tax_rate_range 
    CHECK (
    tax_rate >= 0
      AND tax_rate <= 1
  )"
`;

exports[`pretty: pretty/constraints-10.sql 1`] = `
"ALTER TABLE school.enrollments 
  ADD CONSTRAINT fk_student_course
    FOREIGN KEY(student_id, course_id)
    REFERENCES school.courses_students (student_id, course_id)"
`;

exports[`pretty: pretty/constraints-11.sql 1`] = `
"CREATE TABLE orders (
  id serial PRIMARY KEY,
  user_id int NOT NULL,
  total numeric(10, 2) CHECK (total > 0),
  status varchar(20) DEFAULT 'pending',
  created_at timestamp DEFAULT now(),
  CONSTRAINT fk_user
    FOREIGN KEY(user_id)
    REFERENCES users (id)
    ON DELETE CASCADE,
  CONSTRAINT unique_user_date 
    UNIQUE (user_id, created_at),
  CONSTRAINT check_status 
    CHECK (status IN ('pending', 'completed', 'cancelled'))
)"
`;

exports[`pretty: pretty/constraints-12.sql 1`] = `
"ALTER TABLE products 
  ADD CONSTRAINT fk_category
    FOREIGN KEY(category_id)
    REFERENCES categories (id)
    ON UPDATE CASCADE
    ON DELETE SET NULL
    DEFERRABLE
    INITIALLY DEFERRED"
`;

exports[`pretty: pretty/constraints-13.sql 1`] = `
"ALTER TABLE products 
  ADD CONSTRAINT check_price 
    CHECK (price > 0)"
`;

exports[`pretty: pretty/constraints-14.sql 1`] = `
"ALTER TABLE users 
  ADD CONSTRAINT unique_email 
    UNIQUE (email)"
`;

exports[`pretty: pretty/constraints-15.sql 1`] = `
"ALTER TABLE school.enrollments 
  ADD CONSTRAINT fk_student_course
    FOREIGN KEY(student_id, course_id)
    REFERENCES school.courses_students (student_id, course_id)"
`;

exports[`pretty: pretty/constraints-16.sql 1`] = `
"ALTER TABLE school.enrollments 
  ADD CONSTRAINT chk_enrollment_date 
    CHECK (
    enrollment_date <= CURRENT_DATE
      AND status IN ('active', 'completed', 'withdrawn')
  )"
`;

exports[`pretty: pretty/constraints-17.sql 1`] = `
"CREATE TABLE school.enrollments (
  student_id int NOT NULL,
  course_id int NOT NULL,
  enrollment_date date NOT NULL,
  status text CHECK (status IN ('active', 'completed', 'withdrawn')),
  CHECK (enrollment_date <= CURRENT_DATE)
)"
`;
